akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -346,9 +349,12 @@ final class Split[T](decision: Split.SplitDecision, p: T ⇒ Boolean) extends Gr
             closeThis(handler, elem)
             handOver(handler)
           } else {
-            // Drain into the void
-            if (substreamCancelled) pull(in)
-            else substreamSource.push(elem)
+            if (substreamCancelled) {
```

Comment:
Yes, this is precisely the case that was discussed in both https://github.com/akka/akka/issues/18010:

> To be clear, the desired behavior is that cancelling the response entity should result in a RST being sent to the server and the entire TCP connection being torn down.

and https://github.com/akka/akka/issues/18203:

> In general, I would expect Sink.cancelled to at least free the connection. With respect to your question whether it should abort the connection or just silently consume the entity, I think that probably it would be nice when we could save sending the response over TCP at all, but I have no idea what that means in terms of HTTP (maybe an error on the server side?). Silently consuming the entity on the other hand would also be unexpected when I cancel something. So I would argue that it should close/abort the connection, without knowing the implications for things like keep-alive or pipelining.

To clarify things a bit: I'm not working in the same company as @mspielberg, but we have a very similar use case and cancelling TCP is precisely what we need when response entity is cancelled (e.g. by running with `Sink.cancelled`) or only consumer only partially (e.g. via `take(n)`).

In our case we have an infinite response stream for consuming data from message queue service. This request also acquires a lock on the server side (this has to do with at least once delivery guarantees), which is automatically released when connection is closed. I wrote a client for this service in akka-http which wraps underlying HTTP protocol and exposes data stream in variety of ways for different use cases, the simplest one being a simple `Source[Message]`. The user may choose to consume only a limited amount of messages from this source, which will translate into entity of an underlying chunked `HttpResponse` being cancelled after only required amount of messages has been consumed. The problem is that current implementation keeps draining chunks  in order to prevent connection stalling, since all requests are occupying the same connection, but since the stream is infinite we end up just reading data into /dev/null consuming excessive resources (not good) and also not releasing the lock on the stream and thus preventing anyone else from accessing it (very-very bad).